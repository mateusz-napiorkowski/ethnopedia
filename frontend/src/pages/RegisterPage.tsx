import React, { useEffect, useState } from "react"
import { Formik, Form, Field, ErrorMessage } from "formik"
import * as Yup from "yup"
import { useNavigate } from "react-router-dom"
import { useRegisterMutation } from "../api/auth"
import { Toast, ToastToggle } from "flowbite-react"
import { HiExclamation } from "react-icons/hi"
import { jwtDecode } from "jwt-decode"
import { JWT, useUser } from "../providers/UserProvider"

const RegisterPage = () => {
    const registerMutation = useRegisterMutation()
    const [showErrorToast, setShowErrorToast] = useState(false)
    const { setUserData } = useUser()

    const validationSchema = Yup.object().shape({
        firstName: Yup.string()
            .required("Imię jest wymagane"),

        username: Yup.string()
            .required("Nazwa użytkownika jest wymagana"),

        password: Yup.string()
            .required("Hasło jest wymagane"),

        confirmPassword: Yup.string()
            .oneOf([Yup.ref("password"), ""], "Hasła muszą być takie same")
            .required("Powtórz swoje hasło"),
    })
    const { mutate: registerUser } = registerMutation

    const navigate = useNavigate()

    useEffect(() => {
        let timer: any
        if (showErrorToast) {
            timer = setTimeout(() => setShowErrorToast(false), 3000)
        }
        return () => clearTimeout(timer)
    }, [showErrorToast])

    return <section className="bg-gray-50 dark:bg-gray-900">
        <div
            className="flex flex-col items-center justify-center px-6 py-8 mx-auto md:h-screen lg:py-0
                dark:border-gray-600">

            {showErrorToast && (
                <div
                    className={`fixed bottom-5 right-5 z-50 ${showErrorToast ? "animate-slide-in" : "animate-fade-out"}`}>
                    <Toast>
                        <div
                            className="inline-flex h-8 w-8 items-center justify-center rounded-lg bg-red-100 text-red-500 dark:bg-red-700 dark:text-red-200">
                            <HiExclamation />
                        </div>
                        <div className="mx-3 text-sm font-normal">Użytkownik już istnieje.</div>
                        <ToastToggle onClick={() => setShowErrorToast(false)} />
                    </Toast>
                </div>
            )}

            <div
                className="w-full bg-white rounded-lg shadow dark:border md:mt-0 sm:max-w-md xl:p-0 dark:bg-gray-800
                    dark:border-gray-600 border border-gray-200">
                <div className="p-6 space-y-4 md:space-y-6 sm:p-8">
                    <h1 className="text-xl font-bold leading-tight tracking-tight text-gray-900 md:text-2xl dark:text-white">
                        Zarejestruj się
                    </h1>
                    <Formik
                        initialValues={{ firstName: "", username: "", password: "", confirmPassword: "" }}
                        validationSchema={validationSchema}
                        onSubmit={(values, { setSubmitting }) => {
                            const { firstName, username, password } = values

                            registerUser({ username, firstName, password }, {
                                onSuccess: (response) => {
                                    const { token } = response.data
                                    localStorage.setItem("token", token)
                                    const decodedToken = jwtDecode<JWT>(token)
                                    setUserData(true, decodedToken.firstName, token, decodedToken.userId)
                                    navigate("/")
                                },
                                onError: (error: any) => {
                                    console.error(error)
                                    setShowErrorToast(true)
                                    setSubmitting(false)
                                },
                            })
                        }}
                    >
                        {({ errors, touched }) => (
                            <Form className="space-y-4 md:space-y-6">
                                <div>
                                    <label htmlFor="firstName"
                                           className="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
                                        Imię
                                    </label>
                                    <Field type="text" name="firstName" id="firstName"
                                           className={`bg-gray-50 border ${errors.firstName && touched.firstName ? "border-red-500 bg-red-50" : "border-gray-300"}
                                              text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600
                                              focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700
                                              dark:border-gray-600 dark:placeholder-gray-400 dark:text-white
                                              dark:focus:ring-blue-500 dark:focus:border-blue-500`}
                                           placeholder="Twoje imię" />
                                    <ErrorMessage name="firstName" component="div" className="text-red-600 text-sm" />
                                </div>
                                <div>
                                    <label htmlFor="username"
                                           className="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
                                        Nazwa użytkownika
                                    </label>
                                    <Field type="username" name="username" id="username"
                                           className={`bg-gray-50 border ${errors.username && touched.username ? "border-red-500 bg-red-50" : "border-gray-300"}
                                                                                  text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600
                                                                                  focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700
                                                                                  dark:border-gray-600 dark:placeholder-gray-400 dark:text-white
                                                                                  dark:focus:ring-blue-500 dark:focus:border-blue-500`}
                                           placeholder="Nazwa użytkownika" />
                                    <ErrorMessage name="username" component="div" className="text-red-600 text-sm" />
                                </div>
                                <div>
                                    <label htmlFor="password"
                                           className="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
                                        Hasło
                                    </label>
                                    <Field type="password" name="password" id="password"
                                           className={`bg-gray-50 border ${errors.password && touched.password ? "border-red-500 bg-red-50" :
                                               "border-gray-300"}
                                                                                  text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600
                                                                                  focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700
                                                                                  dark:border-gray-600 dark:placeholder-gray-400 dark:text-white
                                                                                  dark:focus:ring-blue-500 dark:focus:border-blue-500`}
                                           placeholder="••••••••" />
                                    <ErrorMessage name="password" component="div"
                                                  className="text-red-600 text-sm" />
                                </div>
                                <div>
                                    <label htmlFor="confirmPassword"
                                           className="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
                                        Powtórz hasło
                                    </label>
                                    <Field type="password" name="confirmPassword" id="confirmPassword"
                                           className={`bg-gray-50 border ${errors.confirmPassword && touched.confirmPassword ? "border-red-500 bg-red-50" : "border-gray-300"}
                                                                      text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600
                                                                      focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700
                                                                      dark:border-gray-600 dark:placeholder-gray-400 dark:text-white
                                                                      dark:focus:ring-blue-500 dark:focus:border-blue-500`}
                                           placeholder="••••••••" />
                                    <ErrorMessage name="confirmPassword" component="div"
                                                  className="text-red-600 text-sm" />
                                </div>
                                <button type="submit"
                                        className="w-full color-button">
                                    Zarejestruj się
                                </button>
                            </Form>
                        )}
                    </Formik>
                    <p className="text-sm font-light text-gray-500 dark:text-gray-400">
                        Masz już konto?{" "}
                        <span className="ml-1 font-medium text-primary-600 hover:underline dark:text-primary-500
                             cursor-pointer" onClick={() => navigate("/login")}>
                                Zaloguj się.
                            </span>
                    </p>
                </div>
            </div>
        </div>
    </section>
}

export default RegisterPage
