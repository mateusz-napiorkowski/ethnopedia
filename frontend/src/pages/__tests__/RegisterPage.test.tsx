import '@testing-library/jest-dom';
import { render } from '@testing-library/react'
import RegisterPage from "../RegisterPage"
import { MemoryRouter, Routes, Route } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { QueryClient, QueryClientProvider } from 'react-query';

const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom') as any,
    useNavigate: () => mockUseNavigate,
}));

const queryClient = new QueryClient();
const user = userEvent.setup()

const renderComponent = () => {
    return render(
        <QueryClientProvider client={queryClient}>
            <MemoryRouter initialEntries={[`/`]}>
                <Routes>
                    <Route path="/" element={<RegisterPage/>}/>
                </Routes>  
            </MemoryRouter>
        </QueryClientProvider>       
    );
};

describe("RegisterPage tests", () => {
    beforeEach(() => {
        queryClient.clear();
        jest.clearAllMocks()
    });

    it("should render register page", () => {           
        const {container} = renderComponent()
        
        expect(container).toMatchSnapshot()
    })

    it("should call useNavigate with path to login page when log in button is clicked", async () => {           
        const {getByText} = renderComponent()
        const goToLoginPageButton = getByText(/zaloguj się/i)

        await user.click(goToLoginPageButton)

        expect(mockUseNavigate).toHaveBeenCalledWith("/login")
    })

    it.each([
        {
            testName: "should show correct error message when first name was not typed in",
            fieldLabels: ["Nazwa użytkownika", "Hasło", "Powtórz hasło"],
            valuesToTypeIn: ["example username", "examplePassword", "examplePassword"],
            errorMessages: ["Imię jest wymagane", "Nazwa użytkownika jest wymagana", "Hasło jest wymagane",
                "Hasła muszą być takie same", "Powtórz swoje hasło"],
            showErrorMessages: [true, false, false, false, false]
        },
        {
            testName: "should show correct error message when username was not typed in",
            fieldLabels: ["Imię", "Hasło", "Powtórz hasło"],
            valuesToTypeIn: ["example firstName", "examplePassword", "examplePassword"],
            errorMessages: ["Imię jest wymagane", "Nazwa użytkownika jest wymagana", "Hasło jest wymagane",
                "Hasła muszą być takie same", "Powtórz swoje hasło"],
            showErrorMessages: [false, true, false, false, false]
        },
        {
            testName: "should show correct error messages when password was not typed in",
            fieldLabels: ["Imię", "Nazwa użytkownika", "Powtórz hasło"],
            valuesToTypeIn: ["example firstName", "example username", "examplePassword"],
            errorMessages: ["Imię jest wymagane", "Nazwa użytkownika jest wymagana", "Hasło jest wymagane",
                "Hasła muszą być takie same", "Powtórz swoje hasło"],
            showErrorMessages: [false, false, true, true, false]
        },
        {
            testName: "should show correct error messages when repeated password was not typed in",
            fieldLabels: ["Imię", "Nazwa użytkownika", "Hasło"],
            valuesToTypeIn: ["example firstName", "example username", "examplePassword"],
            errorMessages: ["Imię jest wymagane", "Nazwa użytkownika jest wymagana", "Hasło jest wymagane",
                "Hasła muszą być takie same", "Powtórz swoje hasło"],
            showErrorMessages: [false, false, false, false, true]
        },
        {
            testName: "should show correct error message when password and repeated password don't match",
            fieldLabels: ["Imię", "Nazwa użytkownika", "Hasło", "Powtórz hasło"],
            valuesToTypeIn: ["example firstName", "example username", "examplePassword", "otherPassword"],
            errorMessages: ["Imię jest wymagane", "Nazwa użytkownika jest wymagana", "Hasło jest wymagane",
                "Hasła muszą być takie same", "Powtórz swoje hasło"],
            showErrorMessages: [false, false, false, true, false]
        },
        {
            testName: "should show correct error messages when no input fields were filled in",
            fieldLabels: [],
            valuesToTypeIn: [],
            errorMessages: ["Imię jest wymagane", "Nazwa użytkownika jest wymagana", "Hasło jest wymagane",
                "Hasła muszą być takie same", "Powtórz swoje hasło"],
            showErrorMessages: [true, true, true, false, true]
        },
      ])('$testName', async ({fieldLabels, valuesToTypeIn, errorMessages, showErrorMessages}) => {
        const {getByLabelText, getByRole, queryByText } = renderComponent()

        for (const [fieldIndex, fieldLabel] of fieldLabels.entries()) {
            const fieldElement = getByLabelText(fieldLabel);
            await user.type(fieldElement, valuesToTypeIn[fieldIndex]);
        }

        const registerButton = getByRole("button", {name: /zarejestruj się/i})
        await user.click(registerButton)

        errorMessages.forEach(async (message, index) => {
            if(showErrorMessages[index]) {
                expect(queryByText(message)).toBeInTheDocument()
            } else {
                expect(queryByText(message)).not.toBeInTheDocument()
            }
        });
    })
})