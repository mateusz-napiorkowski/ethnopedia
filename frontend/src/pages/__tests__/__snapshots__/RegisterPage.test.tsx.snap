// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterPage tests should render register page 1`] = `
<div>
  <section
    class="bg-gray-50 dark:bg-gray-900"
  >
    <div
      class="flex flex-col items-center justify-center px-6 py-8 mx-auto md:h-screen lg:py-0
                dark:border-gray-600"
    >
      <div
        class="w-full bg-white rounded-lg shadow dark:border md:mt-0 sm:max-w-md xl:p-0 dark:bg-gray-800
                    dark:border-gray-600 border border-gray-200"
      >
        <div
          class="p-6 space-y-4 md:space-y-6 sm:p-8"
        >
          <h1
            class="text-xl font-bold leading-tight tracking-tight text-gray-900 md:text-2xl dark:text-white"
          >
            Zarejestruj się
          </h1>
          <form
            action="#"
            class="space-y-4 md:space-y-6"
          >
            <div>
              <label
                class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
                for="firstName"
              >
                Imię
              </label>
              <input
                class="bg-gray-50 border border-gray-300
                                              text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600
                                              focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700
                                              dark:border-gray-600 dark:placeholder-gray-400 dark:text-white
                                              dark:focus:ring-blue-500 dark:focus:border-blue-500"
                id="firstName"
                name="firstName"
                placeholder="Twoje imię"
                type="text"
                value=""
              />
            </div>
            <div>
              <label
                class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
                for="username"
              >
                Nazwa użytkownika
              </label>
              <input
                class="bg-gray-50 border border-gray-300
                                                                                  text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600
                                                                                  focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700
                                                                                  dark:border-gray-600 dark:placeholder-gray-400 dark:text-white
                                                                                  dark:focus:ring-blue-500 dark:focus:border-blue-500"
                id="username"
                name="username"
                placeholder="Nazwa użytkownika"
                type="username"
                value=""
              />
            </div>
            <div>
              <label
                class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
                for="password"
              >
                Hasło
              </label>
              <input
                class="bg-gray-50 border border-gray-300
                                                                                  text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600
                                                                                  focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700
                                                                                  dark:border-gray-600 dark:placeholder-gray-400 dark:text-white
                                                                                  dark:focus:ring-blue-500 dark:focus:border-blue-500"
                id="password"
                name="password"
                placeholder="••••••••"
                type="password"
                value=""
              />
            </div>
            <div>
              <label
                class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
                for="confirmPassword"
              >
                Powtórz hasło
              </label>
              <input
                class="bg-gray-50 border border-gray-300
                                                                      text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600
                                                                      focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700
                                                                      dark:border-gray-600 dark:placeholder-gray-400 dark:text-white
                                                                      dark:focus:ring-blue-500 dark:focus:border-blue-500"
                id="confirmPassword"
                name="confirmPassword"
                placeholder="••••••••"
                type="password"
                value=""
              />
            </div>
            <button
              class="w-full color-button"
              type="submit"
            >
              Zarejestruj się
            </button>
          </form>
          <p
            class="text-sm font-light text-gray-500 dark:text-gray-400"
          >
            Masz już konto?
             
            <span
              class="ml-1 font-medium text-primary-600 hover:underline dark:text-primary-500
                             cursor-pointer"
            >
              Zaloguj się.
            </span>
          </p>
        </div>
      </div>
    </div>
  </section>
</div>
`;
